### PR TITLE
[#7] [Backend] As a user, I can receive and verify an verification e-mail

### LINKED ISSUE #7: [Backend] As a user, I can receive and verify an verification e-mail
## Why

After the user has signed up successfully, send an e-mail to the user to verify it.

## Acceptance Criteria

- When a user signs up successfully, the user will receive a verification e-mail (with help of Devise)
  - For more convenient testing, use https://github.com/ryanb/letter_opener during development
- The e-mail must include the verification link (this will require a token from Devise)
- Keep the UI of the e-mail as simple as possible!
- Make sure the user can't sign in before the user has verified his e-mail
- Write unit and integration tests

**Note**: Make use of `action_mailer` gem

## Design
N/A

## Resources

https://rubygems.org/gems/actionmailer/versions/5.0.0.1

### PR DESCRIPTION
**Pull request by toby-thanathip**
Close #7

## What happened 👀

- Send verification e-mail after sign-up (with [letter_opener](https://github.com/ryanb/letter_opener) for development, otherwise with [mail-gun](https://github.com/mailgun/mailgun-ruby))
- Require that the e-mail has been verified before the user can sign in (with the help of `user.confirmed?`)
- Navigate to root when the confirmation link has been clicked

## Insight 📝

I had to add `include ActionController::Flash` in `ConfirmationsController` in order for the test to succeed 💭 

Seems like Devise is expecting this when the user is being redirected after confirmation 🤔 

**Note**: Since we don't have `staging.rb`, it will fallback to `production.rb`

## Proof Of Work 📹

**DEVELOPMENT (letter-opener):**

https://github.com/nimblehq/ic-rails-huey-toby/assets/18277915/98562eab-b988-4697-9c55-8a5577e0fe42

**STAGING/PRODUCTION (mail-gun):**

https://github.com/nimblehq/ic-rails-huey-toby/assets/18277915/79d49452-3422-4431-8a1e-206780eaa088

### CODE REVIEW

**Comment by hoangnguyen92dn** [comment]
@Tuubz This PR should belongs to #7 🙈

**Comment by Tuubz** [comment]
@hoangnguyen92dn Hahah, sorryyy. Updated ✅

**Review by olivierobert — APPROVED** [review]

**Comment by Tuubz** [comment]
@hoangnguyen92dn @hoangmirs Resolved conflicts 🚀

**Review by hoangnguyen92dn — COMMENTED** [review]
I am just curious that we don't need to setup email sender to send email to user's mailbox? 🤔

**Comment by Tuubz** [comment]
> I am just curious that we don't need to setup email sender to send email to user's mailbox? 🤔

You mean sending a real e-mail (for staging/production), right? Good question, I didn't thank about 💭

**Comment by Tuubz** [comment]
Sorry guys, I missed this requirement 🙈 I'll put this PR on draft for now and start working on it 🙏

**Review by hoangnguyen92dn — APPROVED** [review]
@Tuubz From the PoW, the confirmation email should be redirected user to staging/production environment url instead of `localhost:3000`, so please don't forget to update the configs on heroku 😄

**Comment by toby-thanathip** [comment]
@hoangnguyen92dn Thank you, I forgot to mention:

- I added the environment variables to staging/production 👍 
- For testing, I changed the values of `development.rb` to test it locally 🙏

File: Gemfile [hunk 1]
```diff
@@ -55,6 +55,7 @@ group :development, :test do
 
   # Utilities
   gem 'figaro' # Simple Rails app configuration
+  gem 'mailgun-ruby' # Contains methods for easily interacting with the Mailgun API
```
Comment (olivierobert) [inline Gemfile]: This should not be in the dev and test environments group only 🙈 This gem would not be installed in the prod environment. It should be moved in the block for `backend` (starting at line 4)

**Review by olivierobert — CHANGES_REQUESTED** [review]

File: app/controllers/api/v1/users/sessions_controller.rb [hunk 2]
```diff
@@ -32,6 +35,20 @@ def render_success(user)
 
           render json: token_data, status: :ok
         end
+
+        def render_invalid_password_error
+          render_error(
+            detail: [I18n.t('devise.failure.invalid', authentication_keys: :email)],
```
Comment (hoangmirs) [inline app/controllers/api/v1/users/sessions_controller.rb]: Sorry, maybe I missed it, but why did we use the array here?

File: spec/requests/sessions_controller_spec.rb [hunk 3]
```diff
@@ -55,6 +55,42 @@
...
+        user = Fabricate(:user, provider: User.providers[:google_oauth2])
+
+        post '/api/v1/users/sign_in', params: {
+          user: { email: user.email, password: user.password },
+          client_id: 'client_id',
+          client_secret: 'client_secret'
+        }
+
+        expect(response).to have_http_status(:unauthorized)
+        expect(JSON.parse(response.body)['errors'][0]['detail']).to include(I18n.t('devise.failure.invalid', authentication_keys: :email))
```
Comment (hoangmirs) [inline spec/requests/sessions_controller_spec.rb]: The detail attribute should be a regular text following `JSON:API` spec:
> a human-readable explanation specific to this occurrence of the problem

And the test can be rewritten as:
```suggestion
        expect(JSON.parse(response.body)['errors'][0]['detail']).to eq(I18n.t('devise.failure.invalid', authentication_keys: :email))
```

**Review by olivierobert — APPROVED** [review]

**Review by hoangmirs — APPROVED** [review]

**Review by hoangmirs — APPROVED** [review]

**Pull request merged**